src/test/java/dog/pawbook/logic/commands/EditDogCommandTest.java

```diff
@@ -0,0 +1,154 @@
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+import dog.pawbook.model.managedentity.owner.Owner;
+import org.junit.jupiter.api.Test;
+
+import dog.pawbook.commons.core.Messages;
+import dog.pawbook.logic.commands.EditDogCommand.EditDogDescriptor;
+import dog.pawbook.model.Database;
+import dog.pawbook.model.Model;
+import dog.pawbook.model.ModelManager;
+import dog.pawbook.model.UserPrefs;
+import dog.pawbook.model.managedentity.Entity;
+import dog.pawbook.model.managedentity.IdMatchPredicate;
+import dog.pawbook.model.managedentity.dog.Dog;
+import dog.pawbook.testutil.DogBuilder;
+import dog.pawbook.testutil.EditDogDescriptorBuilder;
+import javafx.util.Pair;
+
+import java.util.HashSet;
+import java.util.Set;
+
+/**
+ * Contains integration tests (interaction with the Model) and unit tests for EditDogCommand.
+ */
+public class EditDogCommandTest {
+
+    private Model model = new ModelManager(getTypicalDatabase(), new UserPrefs());
+
+    @Test
+    public void execute_allFieldsSpecified_success() {
+        Pair<Integer, Entity> firstIdEntity = model.getDatabase().getEntityList().get(1);
+        Dog editedDog = new DogBuilder((Dog) firstIdEntity.getValue()).build();
+        EditDogCommand.EditDogDescriptor descriptor = new EditDogDescriptorBuilder(editedDog).build();
+        EditDogCommand editDogCommand = new EditDogCommand(firstIdEntity.getKey(), descriptor);
+
+        String expectedMessage = String.format(EditDogCommand.MESSAGE_EDIT_DOG_SUCCESS, editedDog);
+
+        Model expectedModel = new ModelManager(new Database(model.getDatabase()), new UserPrefs());
+        expectedModel.setEntity(firstIdEntity.getKey(), editedDog);
+        expectedModel.updateFilteredEntityList(new IdMatchPredicate(firstIdEntity.getKey()));
+
+        assertCommandSuccess(editDogCommand, model, expectedMessage, expectedModel);
+    }
+
+    @Test
+    public void execute_ownerIdSpecified_success() {
+        Dog toEditDog = (Dog) model.getEntity(ID_TWO);
+
+        DogBuilder dogInList = new DogBuilder(toEditDog);
+        Dog editedDog = dogInList.withName(VALID_NAME_BELL).withBreed(VALID_BREED_BELL)
+                .withTags(VALID_TAG_FRIENDLY).build();
+
+        EditDogCommand.EditDogDescriptor descriptor = new EditDogDescriptorBuilder().withName(VALID_NAME_BELL)
+                .withBreed(VALID_BREED_BELL).withTags(VALID_TAG_FRIENDLY).build();
+        EditDogCommand editEntityCommand = new EditDogCommand(ID_TWO, descriptor);
+
+        String expectedMessage = String.format(EditDogCommand.MESSAGE_EDIT_DOG_SUCCESS, editedDog);
+
+        Model expectedModel = new ModelManager(new Database(model.getDatabase()), new UserPrefs());
+        expectedModel.setEntity(ID_TWO, editedDog);
+        expectedModel.updateFilteredEntityList(new IdMatchPredicate(ID_TWO));
+
+        assertCommandSuccess(editEntityCommand, model, expectedMessage, expectedModel);
+    }
+
+    /*@Test
+    public void execute_someFieldsSpecified_success() {
```

Comment:
Just in case you don't understand how Pawbook works, I will explain it here in the interest of everyone's time. Every dog has an owner, and so long as the dog exists, the owner must also exist, but not the other way around, so owners can be dogless and still exist. While we don't allow for dogs to be created without a valid owner, it is possible to change the ownership afterwards. If the edit requests for dog with ID 2 to be owned by owner with ID 3, obviously the original owner with ID 1 now needs to disown the dog, followed by owner with ID 3 taking over the dog instead. In case you're still confused, 3 entities have been edited at this point, 2 owners 1 dog.

Reply:
Um, I know how Pawbook works lol. Instead of explaining how the function work, can you specify which part the error is caused? I tried adding dogId to new owner and remove dogId from the old owner before I pushed the latest commit here but it doesn't work as well. I'll push the one I did which follows the approach u mentioned here, but it has some assertion error which highlights that the owner to remove dog id doesn't have that dog id. You can have a look at it.

Reply:
First mistake: you copied code from the Class you're testing against mixed in with code of your own, which also contains helper methods that triggered the assertion errors
Second mistake: you have been performing the changes on both the expectedModel and the original model, not sure what you're expecting out of this

Reply:
> can you specify which part the error is caused?

I have already pointed out the second mistake in another comment, not sure what you meant

Reply:
Done!